File: research/uvcc/uvcc-party/uvcc_party/crc32c.py
from __future__ import annotations

"""
CRC32C (Castagnoli) implementation.

UVCC uses CRC32C in some wire formats (optional but deterministic). This module
implements CRC32C with the Castagnoli polynomial (0x1EDC6F41) in pure Python
for portability (CPU and GPU hosts without extra deps).
"""

from functools import lru_cache


def _crc32c_table() -> list[int]:
    # Reflected CRC32C polynomial (Castagnoli): 0x82F63B78
    poly = 0x82F63B78
    tbl: list[int] = []
    for i in range(256):
        c = i
        for _ in range(8):
            if c & 1:
                c = (c >> 1) ^ poly
            else:
                c >>= 1
        tbl.append(c & 0xFFFFFFFF)
    return tbl


@lru_cache(maxsize=1)
def _tbl() -> tuple[int, ...]:
    return tuple(_crc32c_table())


def crc32c(data: bytes, *, init: int = 0) -> int:
    """
    Compute CRC32C(data) with initial value `init` (default 0).

    Returns an unsigned 32-bit integer.
    """
    c = int(init) ^ 0xFFFFFFFF
    tbl = _tbl()
    for b in data:
        c = tbl[(c ^ b) & 0xFF] ^ (c >> 8)
    return (c ^ 0xFFFFFFFF) & 0xFFFFFFFF

File: research/uvcc_native/uvcc/crc32c.cpp
#include "uvcc/crc32c.h"

#include <array>

namespace uvcc {
namespace {

// Reflected CRC32C polynomial (Castagnoli) 0x82F63B78.
constexpr std::uint32_t POLY = 0x82F63B78u;

constexpr std::array<std::uint32_t, 256> make_table() {
    std::array<std::uint32_t, 256> tbl{};
    for (std::uint32_t i = 0; i < 256; i++) {
        std::uint32_t c = i;
        for (int k = 0; k < 8; k++) {
            c = (c & 1u) ? (POLY ^ (c >> 1)) : (c >> 1);
        }
        tbl[i] = c;
    }
    return tbl;
}

constexpr auto TABLE = make_table();

}  // namespace

std::uint32_t crc32c(const void* data, std::size_t len) {
    const auto* p = static_cast<const std::uint8_t*>(data);
    std::uint32_t crc = 0xFFFFFFFFu;
    for (std::size_t i = 0; i < len; i++) {
        crc = TABLE[(crc ^ p[i]) & 0xFFu] ^ (crc >> 8);
    }
    return crc ^ 0xFFFFFFFFu;
}

}  // namespace uvcc


